src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+            {
+                removedDocumentStatesBuilder.Add(getExistingTextDocumentState(oldProjectState, documentId));
+            }
+
+            var removedDocumentStatesForProject = removedDocumentStatesBuilder.ToImmutableAndFree();
+
+            var (newProjectState, compilationTranslationAction) = removeDocumentsFromProjectState(oldProjectState, documentIdsInProject.ToImmutableArray(), removedDocumentStatesForProject);
+
+            (var newSolutionState, _, _) = newCompilationState.Solution.ForkProject(
+                oldProjectState,
+                newProjectState,
+                // Intentionally using this.Solution here and not newSolutionState
+                newFilePathToDocumentIdsMap: this.Solution.CreateFilePathToDocumentIdsMapWithRemovedDocuments(removedDocumentStatesForProject));
+
+            newCompilationState = newCompilationState.ForkProject(
+                newSolutionState,
+                newProjectState,
+                compilationTranslationAction,
+                forkTracker: true);
+        }
+
+        return newCompilationState;
+    }
+
+    /// <inheritdoc cref="Solution.WithCachedSourceGeneratorState(ProjectId, Project)"/>
+    public SolutionCompilationState WithCachedSourceGeneratorStateWorker(ProjectId projectToUpdate, Project projectWithCachedGeneratorState)
+    {
+        this.Solution.CheckContainsProject(projectToUpdate);
+
+        // First see if we have a generator driver that we can get from the other project.
+
+        if (!projectWithCachedGeneratorState.Solution.CompilationState.TryGetCompilationTracker(projectWithCachedGeneratorState.Id, out var tracker) ||
+            tracker.GeneratorDriver is null)
+        {
+            // We don't actually have any state at all, so no change.
+            return this;
+        }
+
+        var projectToUpdateState = this.Solution.GetRequiredProjectState(projectToUpdate);
+
+        // Note: we have to force this fork to happen as the actual solution-state object is not changing.
+        var newCompilationState = this.ForceForkProject(
+            this.Solution,
+            projectToUpdateState,
+            translate: new SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction.ReplaceGeneratorDriverAction(
+                tracker.GeneratorDriver,
+                newProjectState: projectToUpdateState),
+            forkTracker: true);
+
+        return newCompilationState;
+    }
+
+    /// <summary>
+    /// Creates a new solution instance with all the documents specified updated to have the same specified text.
+    /// </summary>
+    public SolutionCompilationState WithDocumentText(IEnumerable<DocumentId?> documentIds, SourceText text, PreservationMode mode)
+    {
+        var result = this;
+
+        foreach (var documentId in documentIds)
+        {
+            if (documentId != null)
+                result = result.WithDocumentText(documentId, text, mode);
+        }
```

Comment:
this was the loop that was simplified.  the original code did work that .WithDocumentText already does.

Reply:
Mea culpa!  Turns out some of the original logic was necessary.  WithDocumentText(DocId) throws if DocId doesn't refer to a real doc in this solution.  But `WithDocumentText(IEnumerable<DocId>)` does not.  Restored the original behavior.